### PR TITLE
Time series models to use sp_to_use

### LINKED ISSUE #1894: Time Series models to use `sp_to_use`
Currently, all models use something like this

```python
self.seasonality_present = globals_dict.get("seasonality_present")
sp = globals_dict.get("seasonal_period")
self.sp = sp if sp is not None else 1
```

```python
def _set_args(self) -> Dict[str, Any]:
    args = {"sp": self.sp} if self.seasonality_present else {}
```

Consider replacing `sp` here with `sp_to_use` instead (from `self.sp_to_use`).

```python
self.sp_to_use = globals_dict.get("sp_to_use")

def _set_args(self) -> Dict[str, Any]:
    args = {"sp": self.sp_to_use}
```

### PR DESCRIPTION
**Pull request by drmario-gh**
## Related Issuse or bug

Closes #1894 

#### Describe the changes you've made

For all time series models that handle seasonality, I make them use `sp_to_use` instead of the `seasonal_period` specified in the experiment setup. 

With this change a couple of None checks were irrelevant, so I removed them as well (indicated with comments on the PR). 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
Unit tests (no new unit tests were added). 

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)

`NA`

## Checklist:
- [ ] My code follows the style guidelines of this project.
Not sure about this one: black pre-commit hook introduced changes in parts of the code I did not modify (in files I modify) and I ended up removing them to avoid mixing them with my changes. Should I reintroduce them in this PR?
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: pycaret/tests/conftest.py [hunk 1]
```diff
@@ -56,6 +56,7 @@ def load_ts_models(load_setup):
         "gpu_param": False,
         "X_train": pd.DataFrame(get_data("airline")),
         "enforce_pi": False,
+        "sp_to_use": 1,
```
Comment (drmario-gh) [inline pycaret/tests/conftest.py:59]: By definition, `sp_to_use` will at least be 1, so I add it in this fixture.

File: pycaret/tests/time_series_test_utils.py [hunk 2]
```diff
@@ -107,6 +107,7 @@ def _return_model_names():
         "X_train": pd.DataFrame(get_data("airline")),
         "enforce_pi": False,
         "seasonal_period": 2,
+        "sp_to_use": 2,
```
Comment (drmario-gh) [inline pycaret/tests/time_series_test_utils.py:110]: And same here, by definition, `sp_to_use` will at least be 1, so I add it in this fixture.

File: pycaret/tests/conftest.py (hunk 1 above)
Comment (ngupta23) [reply pycaret/tests/conftest.py:59]: Will it not be 12 in this case? Since data is monthly seasonal (airline dataset)?

File: pycaret/internal/plots/time_series.py [hunk 3]
```diff
@@ -812,18 +812,12 @@ def plot_time_series_decomposition(
     decomp_result = None
     data_ = data.to_timestamp() if isinstance(data.index, pd.PeriodIndex) else data
 
-    sp_to_use = data_kwargs.get("sp_to_use", None)
+    sp_to_use = data_kwargs.get("sp_to_use")
     if plot == "decomp_classical":
-        if sp_to_use is None:
-            decomp_result = seasonal_decompose(data_, model=classical_decomp_type)
-        else:
             decomp_result = seasonal_decompose(
```
Comment (ngupta23) [inline pycaret/internal/plots/time_series.py:817]: Could you please confirm that this works when `sp_to_use` = 1. If so, then this looks good. Thanks!

File: pycaret/internal/plots/time_series.py [hunk 4]
```diff
@@ -812,18 +812,12 @@ def plot_time_series_decomposition(
...
                 data_, period=sp_to_use, model=classical_decomp_type
             )
     elif plot == "decomp_stl":
-        if sp_to_use is None:
-            decomp_result = STL(data_).fit()
-        else:
             decomp_result = STL(data_, period=sp_to_use).fit()
```
Comment (ngupta23) [inline pycaret/internal/plots/time_series.py:821]: Could you please confirm that this works when `sp_to_use = 1`. If so, then this looks good. Thanks!

**Comment by ngupta23** [comment]
Looks good. Just left some minor comments to check. Will merge after you confirm and all tests pass.

File: pycaret/tests/conftest.py (hunk 1 above)
Comment (drmario-gh) [reply pycaret/tests/conftest.py:59]: 12 definitely makes sense, because for the `airline` dataset, if we don't specify a `seasonal_period`, that's the seasonality that would be inferred from the data and the value that `sp_to_use` will end up having. But that would change the tests as they were before my change and I didn't wanted to introduce such change silently. 

Before my change, all models had:
https://github.com/pycaret/pycaret/blob/e153069683166a2096567e5e4b01df4132e4ad65/pycaret/containers/models/time_series.py#L513-L514
which means that `self.sp` ended up being 1 in the tests.

Other than that, I see no problem in changing the value. Let me know what you think.

File: pycaret/internal/plots/time_series.py (hunk 3 above)
Comment (drmario-gh) [reply pycaret/internal/plots/time_series.py:817]: The thing here is that `sp_to_use` cannot be None. So L818 was unreachable code really.

`seasonal_decompose` does work when `sp_to_use`, although the result is useless: a classical decomposition with no seasonality is all trend. Here is the classical decomposition for the airline data, if we impose `seasonal_period=1` and thus `sp_to_use=1``: 
classical_sp_1

File: pycaret/internal/plots/time_series.py (hunk 4 above)
Comment (drmario-gh) [reply pycaret/internal/plots/time_series.py:821]: As in the classical decomposition, since `sp_to_use` cannot be None, L825 was unreachable code really. But in this case, the STL decomposition does not work with `sp_to_use = 1`: statsmodels code does not allow it (and it makes sense, because it would be pointless otherwise):
```
statsmodels/tsa/_stl.pyx in statsmodels.tsa._stl.STL.__init__()
ValueError: period must be a positive integer >= 2
```
Note also that this is already happening, it is no new behavior. I'd leave this behavior as is: it makes sense for this to fail if `sp_to_use = 1` and the error is pretty informative.

**Comment by drmario-gh** [comment]
> Looks good. Just left some minor comments to check. Will merge after you confirm and all tests pass.

Great. Answered your questions and I just have a small doubt on your first comment.

File: pycaret/tests/conftest.py (hunk 1 above)
Comment (ngupta23) [reply pycaret/tests/conftest.py:59]: But `sp_to_use` would have been 12 before the change as well (not 1). So before the change, `self.sp` (in the model container) would end up taking a value of 12 also (not 1). Did I miss something here?

image

File: pycaret/internal/plots/time_series.py (hunk 4 above)
Comment (ngupta23) [reply pycaret/internal/plots/time_series.py:821]: Fair enough. This is a bug though and we will need to fix this. The appropriate action would be to through a warning message in this case from pycaret rather than have the program crash. I will open a new issue for this.

Thanks for catching this!

File: pycaret/tests/time_series_test_utils.py (hunk 2 above)
Comment (ngupta23) [reply pycaret/tests/time_series_test_utils.py:110]: Did you mean `sp_to_use` will be 2 and not 1?

File: pycaret/tests/conftest.py (hunk 1 above)
Comment (ngupta23) [reply pycaret/tests/conftest.py:59]: You can confirm this by creating an ARIMA model (using `time_series` branch which does not have your changes yet). You will see that it will take sp = 12 in the model.

Comment (drmario-gh) [reply pycaret/tests/conftest.py]: Yes, this one is a bit tricky. I think we are both right but talking about different things.
* Yes, following the regular execution path, if you don't specify `seasonal_period` in the setup, `sp_to_use` is inferred from the data, and ends up being 12.
* However, if we take a look at one of the tests that fail if I don't add this `sp_to_use=1`:
https://github.com/pycaret/pycaret/blob/b1502037b87d035b5d75585a31eb1ac0e9cdce7e/pycaret/tests/test_time_series_blending.py#L43
it uses the fixture `load_models`, which in turn calls `get_all_model_containers`, triggering the init method of all time series models, and it does so with a `globals_dict` that cannot really happen in a regular code execution path: it has no `sp_to_use`.  
https://github.com/pycaret/pycaret/blob/b1502037b87d035b5d75585a31eb1ac0e9cdce7e/pycaret/tests/conftest.py#L78-L86
Thus, before my changes, `self.sp` ends up being 1 in the models initialization:
https://github.com/pycaret/pycaret/blob/b1502037b87d035b5d75585a31eb1ac0e9cdce7e/pycaret/containers/models/time_series.py#L399-L400
After my changes, however, models initialization don't accept a `self.sp=None` and fail if I don't specify an `sp_to_use` in `load_models``:
https://github.com/pycaret/pycaret/blob/e153069683166a2096567e5e4b01df4132e4ad65/pycaret/containers/models/time_series.py#L399
https://github.com/pycaret/pycaret/blob/e153069683166a2096567e5e4b01df4132e4ad65/pycaret/containers/models/time_series.py#L434

This is inconsequential really because in the `load_models` fixture, `load_setup` is called afterwards, so the regular setup method is called and models are finally created through `create_model`, having the right `sp_to_use` which will be 12, no matter the value specified in `load_models`.

https://github.com/pycaret/pycaret/blob/b1502037b87d035b5d75585a31eb1ac0e9cdce7e/pycaret/tests/conftest.py#L65-L72

So tl;dr, it really does not matter the value that we set up here, it is only used in the models init triggered by `get_all_model_containers` , and the call `to `get_all_model_containers` is only used to get all model ids. These ids are then used to create proper model instances which will have the right `sp_to_use=12` value. But what will make more sense probably is directly specifying here `sp_to_use=12`. I'm changing the value to 12.

File: pycaret/tests/time_series_test_utils.py (hunk 2 above)
Comment (drmario-gh) [reply pycaret/tests/time_series_test_utils.py:110]: Yes, sorry, what I wanted to say really is that `sp_to_use` depends on whether the seasonality check for `seasonal_period=2` is positive (in which case `sp_to_use=2`) or negative (in which case `sp_to_use=1`).

In this case, we know that such check is positive and then `sp_to_use=2`.

Comment (ngupta23) [reply pycaret/tests/time_series_test_utils.py:110]: ok, thanks!

File: pycaret/tests/conftest.py (hunk 1 above)
Comment (ngupta23) [reply pycaret/tests/conftest.py]: Thank you for the detailed analysis, clarifying, and changing it to 12. I think at a high level it sounds reasonable, but the only issue might be that when `sp_to_use = 1`, some models like `snaive` (seasonal naive) are disabled. But `sp_to_use = 12` (value > 1) might be safer for that reason.
